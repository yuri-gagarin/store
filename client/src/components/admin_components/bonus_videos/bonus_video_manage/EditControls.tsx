import React from "react";
import { Button, Grid } from "semantic-ui-react";

interface Props {
  handleBonusVideoEdit(e:  React.MouseEvent<HTMLButtonElement>): void;
}

const CancelEditControls: React.FC<Props> = ({ handleBonusVideoEdit }): JSX.Element => {
  return (
    <Grid.Row>
      <Grid.Column>
         <Button content="Close Edit" onClick={handleBonusVideoEdit}></Button>
      </Grid.Column>
    </Grid.Row>
  );
};

export default CancelEditControls;
