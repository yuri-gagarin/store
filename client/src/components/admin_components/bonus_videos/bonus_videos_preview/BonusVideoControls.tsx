import React from "react";
import { } from "semantic-ui-react";
// css imports //
import "./css/bonusVideoControls.css";
// additional components //

interface Props {
  totalBonusVideos: number;
}
const BonusVideossControls: React.FC<Props> = ({ totalBonusVideos }): JSX.Element => {

  return (
    <div className="adminBonusVideosControlsHolder">
      <div className="adminBonusVideosControlsTitle">BonusVideos Controls</div>
      <div className="adminBonusVideosControls">
        <div className="adminBonusVideosCounter">
          <div>Total BonusVideos:</div>
          <div className="adminBonusVideosCount">{totalBonusVideos}</div>
        </div>
      </div>
    </div>
  );
};

export default BonusVideossControls;