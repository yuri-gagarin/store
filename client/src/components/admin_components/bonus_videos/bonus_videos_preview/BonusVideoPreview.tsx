import React from "react";
import { Button, Embed, Item } from "semantic-ui-react";
// css //
import "./css/bonusVideoPreview.css";
// helpers //
import { ConvertDate, setDefaultImage, trimString } from "../../../helpers/displayHelpers";

interface Props {
  bonusVideo: IBonusVideoData;
}

const BonusVideoPreview: React.FC<Props> = ({ bonusVideo }): JSX.Element => {
  const { description, vimeoURL, youTubeURL, createdAt } = bonusVideo;

  if (youTubeURL) {
    return (
      <div className="youtubeBonusVideo">
        <Embed
          url={bonusVideo.youTubeURL}
        />
        <div className="bonusVideoDesc">
          {description}
        </div>
      </div>
    );
  } else if (vimeoURL) {
    return (
      <div>
        <Embed 
          url={bonusVideo.vimeoURL}
        />
        <div className="bonusVideoDesc">
          {description}
        </div>
      </div>
    );
  } else {
    return (
      <div className="bonusVideoErrorDiv">
        <div className="bonusVideoErrorDivMsg">Can't Resolve Video Link</div>
      </div>
    )
  }
};

export default BonusVideoPreview;