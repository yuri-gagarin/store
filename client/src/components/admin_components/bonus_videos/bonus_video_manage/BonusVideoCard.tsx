import React, { useState } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import EditControls from "./EditControls"
// css imports //
import "./css/bonusVideoCard.css";
// actions and state /
import { deleteBonusVideo } from "../actions/APIBonusVideoActions";
import { setCurrentBonusVideo, clearCurrentBonusVideo } from "../actions/UIBonusVideoActions";
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps } from "react-router-dom";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";

interface Props extends RouteComponentProps {
  bonusVideo: IBonusVideoData;
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const BonusVideoCard: React.FC<Props> = ({ bonusVideo, state, dispatch, history }): JSX.Element => {
  const [ editing, setEditing ] = useState<boolean>(false);
  const baseUrl = "/admin/home/my_bonus_videos/manage"
  const { _id, description, youTubeURL, vimeoURL, createdAt, editedAt } = bonusVideo;

  const handleBonusVideoOpen = (e: React.MouseEvent<HTMLButtonElement>): void => {
    history.push(baseUrl + "/view");
  }
  const handleBonusVideoEdit = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (!editing) {
      setCurrentBonusVideo(_id, dispatch, state);
      history.push(baseUrl + "/edit");
      setEditing(true);
    } else {
      clearCurrentBonusVideo(dispatch);
      history.push(baseUrl);
      setEditing(false);
    }
  }
  const handleBonusVideoDelete = (e: React.MouseEvent<HTMLButtonElement>): void => {
    deleteBonusVideo( _id, dispatch, state)
      .then((success) => {
        
      })
  }

  return (
    <React.Fragment>
      <Grid.Row style={{ padding: "0.5em", marginTop: "1em" }}>
        <div className="bonusVideoManageDesc">
          <p><strong>Description:</strong> {description} </p>
          <p><strong>YouTube URL:</strong> {youTubeURL} </p>
          <p><strong>Vimeo URL:</strong> {vimeoURL} </p>
          <p><span>Uploaded At: </span>{ConvertDate.international(createdAt)}</p>
        </div> 
        <div className="bonusVideoManageCtrls">
          <Button 
            inverted
            color="green"
            className="bonusVideoCardBtn" 
            content="Open" 
            onClick={handleBonusVideoOpen} 
          />
          <Button 
            inverted
            color="orange"
            className="bonusVideoCardBtn" 
            content="Edit" 
            onClick={handleBonusVideoEdit}  
          />
          <Button 
            inverted
            color="red"
            className="bonusVideoCardBtn" 
            content="Delete" 
            onClick={handleBonusVideoDelete}
          />
        </div> 
      </Grid.Row>
      {  editing ?  <EditControls handleBonusVideoEdit={handleBonusVideoEdit}/>: null }
    </React.Fragment>
    
  );
};

export default withRouter(BonusVideoCard);