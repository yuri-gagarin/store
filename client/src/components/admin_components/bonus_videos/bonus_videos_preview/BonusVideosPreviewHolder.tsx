import React, { useEffect } from "react";
import { Grid, Item } from "semantic-ui-react";
// css imports //
import "./css/bonusVideosPreviewHolder.css";
// additional components //
import BonusVideoPreview from "./BonusVideoPreview";
import BonusVideosControls from "./BonusVideoControls";
// types and interfaces //
import { AppAction, IGlobalAppState } from "../../../../state/Store";
// api actions //
import { getAllBonusVideos } from "../actions/APIBonusVideoActions";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const BonusVideosPreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { loadedBonusVideos } = state.bonusVideoState;
  useEffect(() => {
    getAllBonusVideos(dispatch);
  }, [dispatch]);

  return (
    <Grid stackable padded columns={2}>
      <Grid.Row>
      <Grid.Column computer={10} tablet={8} mobile={16}>
        <Item.Group>
          {
            loadedBonusVideos.map((video) => {
              return (
                <BonusVideoPreview 
                  key={video._id}
                  bonusVideo={video}
                />
              );
            })
          }
        </Item.Group>
      </Grid.Column>
      <Grid.Column computer={6} tablet={8} mobile={16}>
        <BonusVideosControls totalBonusVideos={loadedBonusVideos.length} />
      </Grid.Column>

      </Grid.Row>
      
    </Grid>
  )
};

export default BonusVideosPreviewHolder;