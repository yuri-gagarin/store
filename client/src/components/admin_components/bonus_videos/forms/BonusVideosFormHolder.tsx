import React, { useState, useEffect, useContext } from "react";
import { Button, Grid } from "semantic-ui-react";
// css imports //
import "./css/bonusVideoFormHolder.css";
// additional components //
import BonusVideoForm from "./BonusVideoForm";
import BonusVideoPreview from "../bonus_videos_preview/BonusVideoPreview";
// state //
import { Store } from "../../../../state/Store";
// api and ui actions //
import { createBonusVideo, editBonusVideo } from "../actions/APIBonusVideoActions";
import { closeBonusVideoForm, openBonusVideoForm } from "../actions/UIBonusVideoActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// type definitions //
import { BonusVideoData, FormState } from "../type_definitions/bonusVideoTypes"
import { RouteComponentProps } from "react-router-dom";

interface Props { }

const BonusVideoFormHolder: React.FC<Props> = ({ }): JSX.Element => {
  const { state, dispatch} = useContext(Store);
  // data for component //
  const { currentBonusVideoData, bonusVideoFormOpen } = state.bonusVideoState;
  const { description, youTubeURL, vimeoURL, createdAt, editedAt } = currentBonusVideoData;

  // form toggle listener //
  const toggleBonusVidform = () => {
    bonusVideoFormOpen ? closeBonusVideoForm(dispatch) : openBonusVideoForm(dispatch);
  }
  // CREATE - EDIT methods //
  const handleCreateBonusVideo = ({ description, youTubeURL, vimeoURL }: FormState): void => {
    const bonusVideoData: BonusVideoData = {
      description,
      youTubeURL,
      vimeoURL
    };

    createBonusVideo(bonusVideoData, dispatch)
      .then((_) => {
        closeBonusVideoForm(dispatch);
      })
      .catch((_) => {
        // handle error show modal ? //
      });
  };
  const handleUpdateBonusVideo = ({ description, youTubeURL, vimeoURL }: FormState): void => {
    const bonusVideoParams: BonusVideoData = {
      description,
      youTubeURL,
      vimeoURL
    };

    editBonusVideo(currentBonusVideoData._id, bonusVideoParams, dispatch, state)
      .then((_) => {
          closeBonusVideoForm(dispatch)
      })
      .catch((_) => {
        // handle an error or show a modal ? //
      })
  };
  // lifecycle hooks //
  useEffect(() => {
    if (!bonusVideoFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ bonusVideoFormOpen ]);
  // component render //
  return (
    <div id="bonusVideoFormHolder">
      {
        checkSetValues<IBonusVideoData>(currentBonusVideoData) ?
          <React.Fragment>
            <Grid.Row id="bonusVideoFormDetails">
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="bonusVideoFormHolderYouTubeUrl">
                  <h3>Video youTube URL:</h3>
                  <p>{youTubeURL}</p>
                </div>
                <div className="bonusVideoFormHolderVimeoUrl">
                  <h3>Video vimeo URL:</h3>
                  <p>{vimeoURL}</p>
                </div>
                <div className="bonusVideoFormHolderDescription">
                  <h3>Video description:</h3>
                  <p>{description}</p>
                </div>
                <div className="bonusVideoFormHolderTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
                <div className="bonusVideoFormPreview">
                  <BonusVideoPreview bonusVideo={currentBonusVideoData} />
                </div>
              </Grid.Column>
            </Grid.Row>
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  
            id="bonusVideoFormToggleBtn" 
            onClick={toggleBonusVidform} 
            content={ bonusVideoFormOpen ? "Close Form" : "Open Form"}
          />
          {
            bonusVideoFormOpen ? 
            <BonusVideoForm 
              description={description} 
              youTubeURL={youTubeURL}
              vimeoURL={vimeoURL}
              handleCreateBonusVideo={handleCreateBonusVideo}
              handleUpdateBonusVideo={handleUpdateBonusVideo}
            /> 
            : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};

export default BonusVideoFormHolder;

