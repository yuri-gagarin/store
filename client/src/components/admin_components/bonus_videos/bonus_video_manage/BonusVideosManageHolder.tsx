import React, { useEffect, useContext, useRef, useState } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import BonusVideoFormHolder from "../forms/BonusVideosFormHolder";
import BonusVideoCard from "./BonusVideoCard";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
import ErrorScreen from "../../miscelaneous/ErrorScreen";
// actions and state //
import { getAllBonusVideos } from "../actions/APIBonusVideoActions";
import { Store } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps, useRouteMatch, Route } from "react-router-dom";

interface Props extends RouteComponentProps {
};

const BonusVideosManageHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, loadedBonusVideos , error } = state.bonusVideoState;
  // local state //
  const [ newDataLoaded, setNewDataLoaded ] = useState<boolean>(false);
  const newBonusVideosRef = useRef(loadedBonusVideos);
  // match route to conditionall render //
  const match = useRouteMatch("/admin/home/my_bonus_videos/manage");

  const handleBack = () => {
    history.goBack();
  };
  // lifycle hooks //
  useEffect(() => {
    let isMounted = true;
    if (isMounted) {
      getAllBonusVideos(dispatch)
        .then((_) => {
          setNewDataLoaded(true);
        })
        .catch((_) => {
          // handle an error ? //
          setNewDataLoaded(false);
        })
    }
    return () => { isMounted = false };
  }, [ dispatch ]); 
  
  useEffect(() => {
    if (newBonusVideosRef.current != loadedBonusVideos && !loading && !error) {
      setNewDataLoaded(true);
    }
  }, [ newBonusVideosRef.current, loadedBonusVideos, loading, error ]);
  // component render //
  return (
    newDataLoaded ?   
    <Grid padded stackable columns={2}>
      <Route path={match?.url + "/edit"}> 
        <Grid.Row>
          <Grid.Column computer={12} tablet={6} mobile={16}>
            <h3>Editing Video: { state.bonusVideoState.currentBonusVideoData.description }</h3>
            <Button inverted color="green" content="Back" onClick={handleBack}></Button>
          </Grid.Column>
        </Grid.Row>
        <BonusVideoFormHolder />
      </Route>
      <Route exact path={match?.url}>
        <Grid.Row>
          <Grid.Column computer={12} tablet={8} mobile={16}>
          {
            loadedBonusVideos.map((bonusVideo) => {
              return (
                <BonusVideoCard 
                  key={bonusVideo._id}
                  bonusVideo={bonusVideo}
                  state={state}
                  dispatch={dispatch}
                />
              );
            })
          }
          </Grid.Column>
          <Grid.Column computer={4} tablet={8} mobile={16}>
            
          </Grid.Column>
        </Grid.Row>
      </Route>
    </Grid>
    :
    (
      error ? <ErrorScreen lastRequest={ () => getAllBonusVideos(dispatch) } /> : <LoadingScreen />
    )
  )
};
// export without router for tests //
export { BonusVideosManageHolder };
// default export //
export default withRouter(BonusVideosManageHolder);