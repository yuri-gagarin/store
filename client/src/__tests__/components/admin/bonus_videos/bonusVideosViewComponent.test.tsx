import React, { ReactElement } from "react";
import { Menu } from "semantic-ui-react";
// testing dependencies //
import { mount, ReactWrapper } from "enzyme";
// component imports //
import BonusVideosViewComponent from "../../../../components/admin_components/bonus_videos/BonusVideosView";
import AdminBonusVideosMenu from "../../../../components/admin_components/menus/AdminBonusVideoMenu";
import { BonusVideosPreviewHolder } from "../../../../components/admin_components/bonus_videos/bonus_videos_preview/BonusVideosPreviewHolder";
import BonusVideosFormHolder from "../../../../components/admin_components/bonus_videos/forms/BonusVideosFormHolder";
import { BonusVideosManageHolder } from "../../../../components/admin_components/bonus_videos/bonus_video_manage/BonusVideosManageHolder";
// additional dependencies //
import  { MemoryRouter as Router, Switch } from "react-router-dom";

describe("BonusVideosView Component test", () => {
  let wrapper: ReactWrapper;
  beforeAll(() => {
    wrapper = mount(
      <Router keyLength={0} initialEntries={["/admin/home/my_videos"]}>
        <BonusVideosViewComponent />
      </Router>
    );
  });
  describe("BonusVideosView Component render test", () => {
    it("Should Properly render BonusVideosView", () => {
      const component = wrapper.find(BonusVideosViewComponent);
      expect(component).toMatchSnapshot();
    });
    it("Should render Admin BonusVideos Menu", () => {
      expect(wrapper.find(AdminBonusVideosMenu)).toHaveLength(1);
    });
    it("Should render conditional routes", () => {
      expect(wrapper.find(Switch)).toHaveLength(1);
    });
    it("Should have Render three children route componets", () => {
      const switchComponent = wrapper.find(Switch);
      expect(switchComponent.props().children).toHaveLength(3);
    });
    it("Should have proper client routes and conditionally render correct Components", () => {
      const switchComponent = wrapper.find(Switch);
      type RouteMap = {
        [key: string]: string
      }
      let map: RouteMap = {};

      const routeComponents = switchComponent.props().children as ReactElement[];
      for (const component of routeComponents) {
        if (component.props.children[1].type.WrappedComponent) {
          // to check if a component is wrapped in a WithRouter() function //
          map[component.props.path as string] = component.props.children[1].type.WrappedComponent;
        } else {
          map[component.props.path as string] = component.props.children[1].type;
        }
      }
      expect(map["/admin/home/my_videos/all"]).toBe(BonusVideosPreviewHolder);
      expect(map["/admin/home/my_videos/create"]).toBe(BonusVideosFormHolder);
      expect(map["/admin/home/my_videos/manage"]).toBe(BonusVideosManageHolder);
    });
  });
  describe("BonusVideosView Component button actions", () => {
    describe("AdminBonusVideosMenu", () => {
      it("Should have 4 main navigation links", () => {
        const adminBonusVideoMenu = wrapper.find(AdminBonusVideosMenu);
        const links = adminBonusVideoMenu.find(Menu.Item)
        expect(links.length).toEqual(4);
      });
      it("'View All Store Items' link should properly function", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = wrapper.find(AdminBonusVideosMenu).find(Menu.Item).at(0);
        viewAllLink.simulate("click");
        // update and assert ///
        wrapper.update();
        expect(wrapper.find(BonusVideosPreviewHolder).length).toEqual(1);
        expect(wrapper.find(BonusVideosFormHolder).length).toEqual(0);
        expect(wrapper.find(BonusVideosManageHolder).length).toEqual(0);
      });
      it("'Create Store Item' link should properly function", () => {
        window.scrollTo = jest.fn();
        const createLink = wrapper.find(AdminBonusVideosMenu).find(Menu.Item).at(1);
        createLink.simulate("click");
        // update and assert //
        wrapper.update();
        expect(wrapper.find(BonusVideosPreviewHolder).length).toEqual(0);
        expect(wrapper.find(BonusVideosFormHolder).length).toEqual(1);
        expect(wrapper.find(BonusVideosManageHolder).length).toEqual(0);
      });
      it("'Manage Store Items' link should properly function", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = wrapper.find(AdminBonusVideosMenu).find(Menu.Item).at(2);
        viewAllLink.simulate("click");
        // update and assert //
        wrapper.update();
        expect(wrapper.find(BonusVideosPreviewHolder).length).toEqual(0);
        expect(wrapper.find(BonusVideosFormHolder).length).toEqual(0);
        expect(wrapper.find(BonusVideosManageHolder).length).toEqual(1);
      });
    });
    
  });
})