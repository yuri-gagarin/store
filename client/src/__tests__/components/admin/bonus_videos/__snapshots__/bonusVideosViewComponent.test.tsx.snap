// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BonusVideosView Component test BonusVideosView Component render test Should Properly render BonusVideosView 1`] = `
<BonusVideosGeneralView>
  <div
    id="adminBonusVideoViewHolder"
  >
    <withRouter(AdminBonusVideosMenu)
      dispatch={[Function]}
    >
      <AdminBonusVideosMenu
        dispatch={[Function]}
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "pathname": "/admin/home/my_videos",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/admin/home/my_videos",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/admin/home/my_videos",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": false,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
      >
        <div
          className="adminBonusVideosMenuFixed"
        >
          <Menu
            className="adminBonusVideosMenu"
            tabular={true}
          >
            <div
              className="ui tabular adminBonusVideosMenu menu"
            >
              <MenuItem
                active={true}
                content="View All Linked Videos"
                name="view_all"
                onClick={[Function]}
              >
                <a
                  className="active item"
                  onClick={[Function]}
                >
                  View All Linked Videos
                </a>
              </MenuItem>
              <MenuItem
                active={false}
                content="Create New Linked Video"
                name="create"
                onClick={[Function]}
              >
                <a
                  className="item"
                  onClick={[Function]}
                >
                  Create New Linked Video
                </a>
              </MenuItem>
              <MenuItem
                active={false}
                content="Manage Linked Videos"
                name="manage"
                onClick={[Function]}
              >
                <a
                  className="item"
                  onClick={[Function]}
                >
                  Manage Linked Videos
                </a>
              </MenuItem>
              <MenuItem
                active={false}
                content="View Sorted Linked Videos"
                name="sorted"
                onClick={[Function]}
              >
                <a
                  className="item"
                  onClick={[Function]}
                >
                  View Sorted Linked Videos
                </a>
              </MenuItem>
            </div>
          </Menu>
        </div>
      </AdminBonusVideosMenu>
    </withRouter(AdminBonusVideosMenu)>
    <Switch />
  </div>
</BonusVideosGeneralView>
`;
